src/subscription/subscribe.js

```diff
@@ -0,0 +1,184 @@
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
+ */
+
+import {
+  addPath,
+  buildExecutionContext,
+  collectFields,
+  defaultFieldResolver,
+  execute,
+  getFieldDef,
+  getOperationRootType,
+} from '../execution/execute';
+import { getArgumentValues } from '../execution/values';
+import { GraphQLSchema } from '../type/schema';
+import invariant from '../jsutils/invariant';
+import mapAsyncIterator from './mapAsyncIterator';
+
+import type {
+  ExecutionContext,
+  ExecutionResult,
+} from '../execution/execute';
+import type {
+  DocumentNode,
+  OperationDefinitionNode,
+} from '../language/ast';
+import type { GraphQLResolveInfo } from '../type/definition';
+
+
+/**
+ * Implements the "Subscribing to request" section of the GraphQL specification.
+ *
+ * Returns an AsyncIterator
+ *
+ * If the arguments to this function do not result in a legal execution context,
+ * a GraphQLError will be thrown immediately explaining the invalid input.
+ */
+export function subscribe(
+  schema: GraphQLSchema,
+  document: DocumentNode,
+  rootValue?: mixed,
+  contextValue?: mixed,
+  variableValues?: ?{[key: string]: mixed},
+  operationName?: ?string,
+): AsyncIterator<ExecutionResult> {
+  // Note: these invariants are identical to execute.js
+  invariant(schema, 'Must provide schema');
+  invariant(document, 'Must provide document');
+  invariant(
+    schema instanceof GraphQLSchema,
+    'Schema must be an instance of GraphQLSchema. Also ensure that there are ' +
+    'not multiple versions of GraphQL installed in your node_modules directory.'
+  );
+
+  // Variables, if provided, must be an object.
+  invariant(
+    !variableValues || typeof variableValues === 'object',
+    'Variables must be provided as an Object where each property is a ' +
+    'variable value. Perhaps look to see if an unparsed JSON string ' +
+    'was provided.'
+  );
+
+  // If a valid context cannot be created due to incorrect arguments,
+  // this will throw an error.
+  const exeContext = buildExecutionContext(
+    schema,
+    document,
+    rootValue,
+    contextValue,
+    variableValues,
+    operationName
+  );
+
+  // Call the `subscribe()` resolver or the default resolver to produce an
+  // AsyncIterable yielding raw payloads.
+  const subscription = resolveSubscription(
+    exeContext,
+    exeContext.operation,
+    rootValue
+  );
+
+  // For each payload yielded from a subscription, map it over the normal
+  // GraphQL `execute` function, with `payload` as the rootValue.
+  return mapAsyncIterator(
+    subscription,
+    payload => execute(
+      schema,
+      document,
+      payload,
+      contextValue,
+      variableValues,
+      operationName
+    )
+  );
+}
+
+function resolveSubscription(
```

Comment:
Sure - we can get to naming concerns. I actually think `resolve` is the correct name in that it's the result of the Resolve algorithm - but perhaps this would be more clear after some refactoring. Since this function is almost entirely recycling of code written in `execute.js`, I'd love for those parts to be more clearly referenced by function call so the purpose of this function is more clear.